### PR TITLE
feat: support replay cancellation & timeout graceful shutdown

### PR DESCRIPTION
**Pull request by arinda-arif**
Feature & Bugfix:
- [x] add replay cancel command
- [x] add replay status checks in replay loop
- [x] fix context deadline exceeded when marking replay as failed (due to timeout)
- [x] show better logs when replay validation not pass

Dependency (needs to be merged first): 
- [x] [proton #66](https://github.com/goto/proton/pull/66)

### CODE REVIEW

File: core/scheduler/service/replay_validator.go [hunk 1]
```diff
@@ -75,7 +71,7 @@ func (v Validator) validateConflictedReplay(ctx context.Context, replayRequest *
 		// Check any intersection of date range
 		if (onGoingReplay.Config().StartTime.Equal(replayRequest.Config().EndTime) || onGoingReplay.Config().StartTime.Before(replayRequest.Config().EndTime)) &&
 			(onGoingReplay.Config().EndTime.Equal(replayRequest.Config().StartTime) || onGoingReplay.Config().EndTime.After(replayRequest.Config().StartTime)) {
-			return errors.NewError(errors.ErrFailedPrecond, scheduler.EntityJobRun, "conflicted replay found")
+			return errors.NewError(errors.ErrFailedPrecond, scheduler.EntityJobRun, fmt.Sprintf("request is conflicted with on going replay with ID %s", onGoingReplay.ID().String()))
```
Comment (Mryashbhardwaj) [inline core/scheduler/service/replay_validator.go:74]:
```suggestion
			return errors.NewError(errors.ErrFailedPrecond, scheduler.EntityJobRun, fmt.Sprintf("request is conflicted with an on going replay with ID %s", onGoingReplay.ID().String()))
```

File: core/scheduler/service/replay_validator.go [hunk 2]
```diff
@@ -93,7 +89,7 @@ func (v Validator) validateConflictedRun(ctx context.Context, replayRequest *sch
 	}
 	for _, run := range runs {
 		if run.State == scheduler.StateQueued || run.State == scheduler.StateRunning {
-			return errors.NewError(errors.ErrFailedPrecond, scheduler.EntityJobRun, "conflicted job run found")
+			return errors.NewError(errors.ErrFailedPrecond, scheduler.EntityJobRun, fmt.Sprintf("conflicted job run found. job run %s is already in queued/running state", run.ScheduledAt.Format(time.DateTime)))
```
Comment (Mryashbhardwaj) [inline core/scheduler/service/replay_validator.go:92]: can we not show the state from `run.State`

File: core/scheduler/service/replay_worker.go [hunk 3]
```diff
@@ -63,8 +64,16 @@ func (w *ReplayWorker) Execute(replayID uuid.UUID, jobTenant tenant.Tenant, jobN
...
+		defer cleanupCancelFn()
+
+		errMessage := err.Error()
+		if errors.Is(err, context.DeadlineExceeded) {
+			errMessage = "replay execution timed out"
+		}
+		w.logger.Error("[ReplayID: %s] unable to execute replay for job [%s]: %s", replayID.String(), jobName.String(), errMessage)
+
+		if err := w.replayRepo.UpdateReplayStatus(cleanupCtx, replayID, scheduler.ReplayStateFailed, errMessage); err != nil {
 			w.logger.Error("[ReplayID: %s] unable to update replay to failed: %s", replayID, err.Error())
```
Comment (Mryashbhardwaj) [inline core/scheduler/service/replay_worker.go:77]:
```suggestion
			w.logger.Error("[ReplayID: %s] unable to set replay status to 'failed': %s", replayID, err.Error())
```

File: core/scheduler/service/replay_worker.go [hunk 4]
```diff
@@ -97,6 +106,11 @@ func (w *ReplayWorker) startExecutionLoop(ctx context.Context, replayID uuid.UUI
 		}
 		replayWithRun := storedReplayWithRun
 
+		if storedReplayWithRun.Replay.IsTerminated() {
+			w.logger.Info("replay [%s] is externally terminated with status [%s]", replayWithRun.Replay.ID().String(), storedReplayWithRun.Replay.State().String())
```
Comment (Mryashbhardwaj) [inline core/scheduler/service/replay_worker.go:110]: i think the logging format is not consistent with https://github.com/goto/optimus/pull/194/files#diff-c7f1cdd7546059d512af583453b54e1ed17f9498daed33ec1b55f96bac96d5dbR74

**Review by Mryashbhardwaj — APPROVED** [review]
looks good, just minor logging / error msgs fixes suggested

File: core/scheduler/service/replay_worker.go (hunk 4 above)
Comment (arinda-arif) [reply core/scheduler/service/replay_worker.go]: updated

File: core/scheduler/service/replay_validator.go (hunk 2 above)
Comment (arinda-arif) [reply core/scheduler/service/replay_validator.go]: updated

**Pull request merged**